File: 0002-add-two-numbers.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        if not all([l1, l2]):
            return l1 or l2

        l1_list = [l1.val]
        head = l1
        while(head.next):
            head = head.next
            l1_list.append(head.val)

        l2_list = [l2.val]
        head = l2
        while(head.next):
            head = head.next
            l2_list.append(head.val)

        l1_list.reverse()
        l2_list.reverse()

        if l1_list[0] == 0 or l2_list[0] == 0:
            return l2 if l1_list[0] == 0 else l1

        # 补 0
        if len(l1_list) > len(l2_list):
            l2_list = [0] * (len(l1_list) - len(l2_list)) + l2_list
        else:
            l1_list = [0] * (len(l2_list) - len(l1_list)) + l1_list

        result_list = []
        add_value = 0
        for i in range(len(l1_list) - 1, -1, -1):
            result_list.append((l1_list[i] + l2_list[i] + add_value) % 10)
            add_value = (l1_list[i] + l2_list[i] + add_value) // 10

        if add_value > 0:
            result_list.append(add_value)

        result = ListNode(result_list[0], None)

        head = result

        for i in result_list[1:]:
            temp = ListNode(i, None)
            head.next = temp
            head = temp
        return result


if __name__ == "__main__":
    l1 = ListNode(1, None)
    l2 = ListNode(0, None)

    head = l1
    for i in [8, 8, 1, 2]:
        temp = ListNode(i, None)
        head.next = temp
        head = temp

    head = l2
    for i in [6, 7, 9, 8, 4]:
        temp = ListNode(i, None)
        head.next = temp
        head = temp

    head = l1
    while head:
        print(head.val, sep=" ", end=" ")
        head = head.next

    print()

    head = l2
    while head:
        print(head.val, sep=" ", end=" ")
        head = head.next

    print()

    head = Solution().addTwoNumbers(l1, l2)
    while head:
        print(head.val, sep=" ", end=" ")
        head = head.next
    print()

File: 0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 */
#include <iostream>

// class ListNode {
//     public: int val;
//     ListNode *next;
//     ListNode() : val(0), next(nullptr) {}
//     ListNode(int x) : val(x), next(nullptr) {}
//     ListNode(int x, ListNode *next) : val(x), next(next) {}
// };

struct ListNode {
     int val;
     ListNode *next;
     ListNode() : val(0), next(nullptr) {}
     ListNode(int x) : val(x), next(nullptr) {}
     ListNode(int x, ListNode *next) : val(x), next(next) {}
 };

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode * result = nullptr, *head = nullptr;
        int add_value = 0, operate_value = 0;
        while (l1 or l2) {
            operate_value = 0;
            if (l1) {
                operate_value += l1->val;
                l1 = l1->next;
            }
            if (l2) {
                operate_value += l2->val;
                l2 = l2->next;
            }
            auto temp = new ListNode((operate_value + add_value) % 10);
            add_value = (operate_value + add_value) / 10;
            if (nullptr == result) {
                // result = (ListNode *)std::malloc(sizeof(ListNode));
                result = temp;
                head = result;
            }else {
                head->next = temp;
                head = head->next;
            }
        }
        if (add_value > 0) {
            auto temp = new ListNode(add_value);
            head->next = temp;
        }
        return result;
    }
};

void print(ListNode * ln) {
    std::cout<<"ListNode: ";
    while (nullptr != ln) {
        std::cout<<ln->val<<" ";
        ln = ln->next;
    }
    std::cout<<std::endl;
}

ListNode * create(ListNode * ln, int *nums, int size){
    ListNode * head = nullptr;
    for (int i=0; i<size; i++) {
        auto temp = new ListNode(*(nums+i));
        if (nullptr == ln) {
            ln = temp;
            head = ln;
        }else {
            head->next = temp;
            head = head->next;
        }
    }
    return ln;
}

int main(int argc, char const *argv[]) {
    ListNode * l1 = nullptr, *l2 = nullptr, *head = nullptr;

    // int nums1[] = {2, 4, 3};
    int nums1[] = {9,9,9,9,9,9,9};

    l1 = create(l1, nums1, sizeof(nums1)/sizeof(int));

    // int nums2[] = {5, 6, 4};
    int nums2[] = {9,9,9,9};

    l2 = create(l2, nums2, sizeof(nums2)/sizeof(int));

    print(l1);
    print(l2);
    print(Solution().addTwoNumbers(l1, l2));
    return 0;
}